deep_qa/data/fields/field.py

```diff
@@ -1,43 +1,53 @@
-"""
-A ``Field`` is some piece of a data instance that ends up as an array in a model (either as an
-input or an output).  Data instances are just collections of fields.
-
-
-Fields go through up to two steps of processing: (1) tokenized fields are converted into token ids,
-(2) fields containing token ids (or any other numeric data) are padded (if necessary) and converted
-into data arrays.  We have two ``Field`` objects corresponding to fields in each stage of this
-processing pipeline: ``UnindexedFields`` are fields containing raw strings (already tokenized) that
-need to be converted into token IDs.  We use these fields to compute a vocabulary, then convert
-them to ``IndexedFields``.  ``IndexedFields`` have a method for determining padding lengths, so
-that a data generator object can intelligently batch together instances, then pad them.
-
-If you are writing a new ``Field`` class, if there is nothing in the field that needs to be
-converted from strings to integers, you should just subclass ``IndexedField``.  If you `do` need to
-convert strings to integers, you should subclass both ``IndexedField`` and ``UnindexedField``, and
-make the ``UnindexedField`` subclass public, and the other private.
-"""
 from typing import Dict, List
 
 import numpy
 
 from ..vocabulary import Vocabulary
 
-# Because of how the dependencies work between these objects, we're defining them in reverse order,
-# so that UnindexedField can reference IndexedField in its return types.
-
-class IndexedField:
+class Field:
     """
-    An ``IndexedField`` is a field that is almost ready to be put into a data array for use in a
-    model.  The only thing missing is padding.  The methods on this object allow you to get padding
-    lengths, so you can group things into batches of similar size, and then actually pad the
-    field into a fixed-size array that can be used in a batch of instances.
+    A ``Field`` is some piece of a data instance that ends up as an array in a model (either as an
+    input or an output).  Data instances are just collections of fields.
+
+    Fields go through up to two steps of processing: (1) tokenized fields are converted into token
+    ids, (2) fields containing token ids (or any other numeric data) are padded (if necessary) and
+    converted into data arrays.  The ``Field`` API has methods around both of these steps,
+    though they may not be needed for some concrete ``Field`` classes - if your field doesn't have
+    any strings that need indexing, you don't need to implement ``count_vocab_items`` or ``index``.
+    These methods ``pass`` by default.
+
+    Once a vocabulary is computed and all fields are indexed, we will determine padding lengths,
+    then intelligently batch together instances and pad them into actual arrays.
     """
+    def count_vocab_items(self, counter: Dict[str, Dict[str, int]]):
```

Comment:
Regardless, I don't think it's desirable to set the vocabulary in this manner.  One use case is pre-trained word vectors.  You typically want to include all tokens in with pre-trained vectors in your task vocabulary, since even if a token isn't in the training data set but is in the test set you'll be able to take advantage of the pre-trained vector.